base/math.jl

```diff
@@ -536,38 +536,48 @@ Stacktrace:
 [...]
 ```
 """
-hypot(x::Number, y::Number) = hypot(promote(x, y)...)
-function hypot(x::T, y::T) where T<:Number
-    ax = abs(x)
-    ay = abs(y)
-    if ax < ay
-        ax, ay = ay, ax
-    end
+function hypot(x::Number, y::Number)
+	ay, ax = minmax(abs(x), abs(y))
     if iszero(ax)
         r = ay / oneunit(ax)
     else
         r = ay / ax
     end
-
     rr = ax * sqrt(1 + r * r)
 
-    # Use type of rr to make sure that return type is the same for
-    # all branches
-    if isnan(r)
-        isinf(ax) && return oftype(rr, Inf)
-        isinf(ay) && return oftype(rr, Inf)
-        return oftype(rr, r)
-    else
-        return rr
-    end
+	if isnan(r)
+		if isinf(x) || isinf(y)
+			return oftype(rr, Inf)
+		else
+			return oftype(rr, NaN)
+		end
+   	else
+   		return rr
+   	end
 end
-
+#= Specialized one and two argument versions are
+faster than generic varargs version below =#
+hypot(x::Number) = abs(float(x))
 """
     hypot(x...)
 
 Compute the hypotenuse ``\\sqrt{\\sum x_i^2}`` avoiding overflow and underflow.
 """
-hypot(x::Number...) = sqrt(sum(abs2(y) for y in x))
+hypot(x::Number, y::Number...) = _hypot(promote(x, y...))
+function _hypot(x::NTuple{N,<:Number}) where {N}
+	# speculatively try fast naive approach
+	s = sum(abs2, float.(x))
+	r = sqrt(s)
+	if isnan(s) || isinf(s) || s ≤ floatmin(s) * (N-1)
+		# if overflow/underflow, try normalization
+		any(isinf, x) && return oftype(r, Inf) # IEEE 754
```

Comment:
@stevengj I think this is what you are suggesting. I also tried `isnan(s) && any(isinf, x) && return oftype(r, Inf)`, but that fails this test: `@test hypot(1s, 2s) ≈ s * hypot(1,2)   rtol=8eps(T)` with `T=Float32`.

Reply:
Shouldn't that be
```
if isnan(s)
    return any(isinf, x) ? oftype(r, Inf) : r
end
```
Seems like this is better for an underflow case like `hypot(0,0,0,0,0,0,0)` so that it avoids checking `isinf` on all the arguments.

Reply:
(Seems better just to write out `sqrt(s)` in several places rather than run the risk of computing it unnecessarily.)

Reply:
Done. But I think we compute `r = sqrt(s)` in all execution paths. Unless `oftype(f(x), y)` is able to elide evaluation of `f(x)` and just infer its return type?